lib/internal/fs/streams.js

```diff
@@ -395,6 +397,14 @@ WriteStream.prototype.close = function(cb) {
 // There is no shutdown() for files.
 WriteStream.prototype.destroySoon = WriteStream.prototype.end;
 
+WriteStream.prototype.fsync = function(callback) {
+  if (callback) {
+    return fsync(this.fd, callback);
+  } else {
+    return fsyncSync(this.fd);
+  }
```

Comment:
It's important that `fsync()` calls are ordered with respect to `write()` calls that come before and after it; this PR currently doesn't do that. Another way to think about that is that `fsync()` is a barrier.
```js
const ws = fs.createWriteStream('test.txt')
ws.write('important data')  // this should hit the disk first
ws.fsync()  // then this
ws.write('more important data')  // then this
ws.fsync()  // then this
```
What your PR does is this:

1. When you pass a callback to `fsync()`: `fsync()` and `write()` race each other. It's indeterminate which one finishes first.

2. When you don't pass a callback: `fsync()` executes before the `write()` that precedes it in program order because the write is async and the `fsync()` call isn't.

Reply:
what about we add a new event named like `writed` and call it after `write()` and register the `fsync`?

Reply:
maybe we refactor the `write` part to event-driven.

every `write` finished and call the next `write` event. that we can inject fsync on it.🤔